src/ott/geometry/geodesic.py

```diff
@@ -0,0 +1,260 @@
+    assert G.shape[0] == G.shape[1], G.shape
+
+    if directed:
+      G = G + G.T
+
+    degree = jnp.sum(G, axis=1)
+    laplacian = jnp.diag(degree) - G
+
+    if normalize:
+      inv_sqrt_deg = jnp.diag(
+          jnp.where(degree > 0.0, 1.0 / jnp.sqrt(degree), 0.0)
+      )
+      laplacian = inv_sqrt_deg @ laplacian @ inv_sqrt_deg
+
+    if t is None:
+      t = (jnp.sum(G) / jnp.sum(G > 0.)) ** 2
+
+    # Create an instance of the Geodesic class and set the attribute
+    geodesic_instance = cls(laplacian, t=t, order=order, **kwargs)
+
+    # Compute the coeffs of the Chebyshev pols approx using Bessel functs.
+    chebyshev_coeffs = (
+        2 *
+        ive(jnp.arange(0, geodesic_instance.order + 1), -geodesic_instance.t)
+    ).tolist()
+
+    # Set the attribute
+    geodesic_instance.chebyshev_coeffs = chebyshev_coeffs
+
+    return geodesic_instance
+
+  def apply_kernel(
+      self,
+      scaling: jnp.ndarray,
+      eps: Optional[float] = None,
+      axis: int = 0,
+  ) -> jnp.ndarray:
+    r"""Apply :attr:`kernel_matrix` on positive scaling vector.
+
+    Args:
+    scaling: Scaling to apply the kernel to.
+    eps: passed for consistency, not used yet.
+    axis: passed for consistency, not used yet.
+
+    Returns:
+    Kernel applied to ``scaling``.
+    """
+
+    def compute_largest_eigenvalue(laplacian_matrix, k, seed=None):
+      # Compute the largest eigenvalue of the Laplacian matrix.
+      if seed is None:
+        seed = jax.random.PRNGKey(0)
+      n, _ = self.shape
+      # Generate random initial directions for eigenvalue computation
+      initial_dirs = jax.random.normal(seed, (n, k))
+
+      # Create a sparse matrix-vector product function using sparsify
+      # This function multiplies the sparse laplacian_matrix with a vector
+      lapl_vector_product = jesp.sparsify(lambda v: laplacian_matrix @ v)
+
+      # Compute eigenvalues using the sparse matrix-vector product
+      eigvals, _, _ = jesp.linalg.lobpcg_standard(
+          lapl_vector_product, initial_dirs, m=k
+      )
+
+      return np.max(eigvals)
+
+    def rescale_laplacian(laplacian_matrix: jnp.ndarray) -> jnp.ndarray:
+      # Rescale the Laplacian matrix.
+      largest_eigenvalue = compute_largest_eigenvalue(laplacian_matrix, k=1)
+      if largest_eigenvalue > 2:
+        rescaled_laplacian = laplacian_matrix.copy()
+        rescaled_laplacian /= largest_eigenvalue
+        return 2 * rescaled_laplacian
+      return laplacian_matrix
+
+    def define_scaled_laplacian(laplacian_matrix: jnp.ndarray) -> jnp.ndarray:
+      # Define the scaled Laplacian matrix.
+      n = laplacian_matrix.shape[0]
+      identity = jnp.eye(n)
+      return laplacian_matrix - identity
+
+    def compute_chebyshev_approximation(
+        x: jnp.ndarray, coeffs: List[float]
+    ) -> jnp.ndarray:
+      # Compute the Chebyshev polynomial approximation for
+      # the given input and coefficients.
+      x_dense = x.toarray() if sp.issparse(x) else x
```

Comment:
Don't use `scipy` here, assume jax.Array or jesp.BCOO

Reply:
Changed it to jesp.BCOO